File: Tutorials/ObjC/Example7.3/Example7.3/ViewController.m
//
//  ViewController.m
//  Example7.3
//
//  Created by Penthera on 3/3/21.
//

#import "ViewController.h"
#import "DemoPlayerViewController.h"
#import "PlaylistViewController.h"

static NSString* PlaylistName = @"TESTQUEUE-8";


// ---------------------------------------------------------------------------------------------------------
// IMPORTANT:
// The following three values must be initialzied, please contact support@penthera.com to obtain these keys
// ---------------------------------------------------------------------------------------------------------
static NSString* backplaneUrl = @"replace_with_your_backplane_url";                                         // <-- change this
static NSString* publicKey = @"replace_with_your_public_key";   // <-- change this
static NSString* privateKey = @"replace_with_your_private_key";  // <-- change this

@interface ViewController () <VirtuosoDownloadEngineNotificationsDelegate>

//
// MARK: Instance data
//
@property (nonatomic, strong) VirtuosoDownloadEngineNotificationManager* downloadEngineNotifications;
@property (nonatomic, strong)NSMutableArray* assets;
@property (nonatomic, strong)VirtuosoPlaylist* playlist;
@property (nonatomic, strong)NSOperationQueue* loadQueue;
@property (nonatomic, assign)Boolean assetsDeleted;
@end


@implementation ViewController

//
// MARK: Lifecycle methods
//

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // download engine update listener
    self.downloadEngineNotifications = [[VirtuosoDownloadEngineNotificationManager alloc]initWithDelegate:self];

    self.navigationItem.leftBarButtonItem =  [[UIBarButtonItem alloc]initWithTitle:@"Reset"
                                                                             style:UIBarButtonItemStyleDone
                                                                            target:self
                                                                            action:@selector(resetClicked)];
    
    self.navigationItem.rightBarButtonItem =  [[UIBarButtonItem alloc]initWithTitle:@"Add"
                                                                             style:UIBarButtonItemStyleDone
                                                                            target:self
                                                                            action:@selector(addClicked)];
    
    self.tableView.rowHeight = UITableViewAutomaticDimension;
    self.tableView.estimatedRowHeight = 80;

    self.loadQueue = [NSOperationQueue new];
    self.loadQueue.maxConcurrentOperationCount = 1;
    self.loadQueue.qualityOfService = NSQualityOfServiceUtility;
    
    self.assets = [NSMutableArray new];
    [self startup];
}

//
// MARK: TableView Delegate and Data
//
-(NSInteger)numberOfSectionsInTableView:(UITableView *)tableView
{
    return 1;
}

-(NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return self.assets.count;
}

-(UITableViewCell*)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    UITableViewCell* cell = [tableView dequeueReusableCellWithIdentifier:@"defaultCell" forIndexPath:indexPath];
    
    VirtuosoAsset* asset = [self.assets objectAtIndex:indexPath.row];
    cell.textLabel.text = asset.assetID;
    
    NSDateFormatter* formatter = [NSDateFormatter new];
    formatter.dateStyle = NSDateFormatterMediumStyle;
    formatter.timeStyle = NSDateFormatterMediumStyle;

    NSMutableString* sb = [NSMutableString new];
    if (asset.fastPlayEnabled)
    {
        [sb appendString:@"FastPlay Enabled\n"];
    }
    if (asset.fastPlayReady)
    {
        [sb appendString:@"FastPlay Ready\n"];
    }
    cell.detailTextLabel.text = sb;

    return cell;
}

-(void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    VirtuosoAsset* asset = [self.assets objectAtIndex:indexPath.row];
    
    if (asset)
    {
        [self playAsset:asset];
    }
    
    [tableView deselectRowAtIndexPath:indexPath animated:TRUE];
}

//
// MARK: Internal methods
//

-(void)startup
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self clearData];
        
        //
        // Enable the Engine
        VirtuosoDownloadEngine.instance.enabled = TRUE;
        
        // Backplane permissions require a unique user-id for the full range of captabilities support to work
        // Production code that needs this will need a unique customer ID.
        // For demonstation purposes only, we use the device name
        NSString* userName = UIDevice.currentDevice.name;
        
        //
        // Create the engine confuration
        VirtuosoEngineConfig* config = [[VirtuosoEngineConfig alloc]initWithUser:userName
                                                                    backplaneUrl:backplaneUrl
                                                                       publicKey:publicKey
                                                                      privateKey:privateKey];
        if (!config)
        {
            dispatch_async(dispatch_get_main_queue(), ^{
                UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Setup Required"
                                                                               message:@"Please contact support@penthera.com to setup the backplaneUrl, publicKey, and privateKey"
                                                                        preferredStyle:UIAlertControllerStyleAlert];
                UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                    exit(0);
                }];
                [alert addAction:ok];
                [self presentViewController:alert animated:TRUE completion:nil];

            });
            return;
        }

        //
        // Start the Engine
        // This method will execute async, the callback will happen on the main-thread.
        [VirtuosoDownloadEngine.instance startup:config startupCallback:^(kVDE_EngineStartupCode status) {
            if (status == kVDE_EngineStartupSuccess) {
                NSLog(@"Startup succeeded.");
                [self createInitialAsset];
            } else {
                NSLog(@"Startup encountered error.");
            }
        }];
    });
}

-(void)resetClicked
{
    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
        [self clearData];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self refreshView];
        });
    });
}

-(void)addClicked
{
    UIStoryboard* storyboard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    PlaylistViewController* view = [storyboard instantiateViewControllerWithIdentifier:@"PlaylistViewController"];
    view.playlist = [VirtuosoPlaylist find:PlaylistName];
    [self.navigationController pushViewController:view animated:true];
    
}

-(void)clearData
{
    NSAssert(![NSThread isMainThread], @"This method should not be called on MainThread");
    
    // Blow away existing Playlist, this needs to happen
    // before we remove the assets.
    [VirtuosoPlaylist clear:PlaylistName];
    
    self.playlist = nil;

    // Remove all assets
    self.assetsDeleted = false;
    [VirtuosoAsset deleteAll];
    
    // Wait for delete to complete
    while(!self.assetsDeleted)
    {
        [NSThread sleepForTimeInterval:1];
    }
    
    // Create demo Playlist
    if (![self initializeData]) {
        dispatch_async(dispatch_get_main_queue(), ^{
            UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Initialization error"
                                                                           message:@"Check logs for reason."
                                                                    preferredStyle:UIAlertControllerStyleAlert];
            UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
                return;
            }];
            [alert addAction:ok];
            [self presentViewController:alert animated:TRUE completion:nil];

        });

        return;
    }
    // create initial asset
    if (VirtuosoDownloadEngine.instance.started) {
        [self createInitialAsset];
    }
    [self loadData];
}

-(void)refreshView
{
    if (self.playlist.items.count < 6)
    {
        self.navigationItem.rightBarButtonItem = [[UIBarButtonItem alloc]initWithTitle:@"Add" style:UIBarButtonItemStyleDone target:self action:@selector(addClicked)];
    }
    else
    {
        self.navigationItem.rightBarButtonItem =  [[UIBarButtonItem alloc]initWithTitle:@"Browse"
                                                                                 style:UIBarButtonItemStyleDone
                                                                                target:self
                                                                                action:@selector(addClicked)];
    }
}

-(Boolean)initializeData
{
    NSError* error = nil;
    
    VirtuosoPlaylistConfig* playlistConfig = [[VirtuosoPlaylistConfig alloc]initWithName:PlaylistName
                                                                            playlistType:kVDE_PlaylistType_FastPlay
                                                                                   error:&error];
    if (error)
    {
        VLog(kVL_LogError, @"Failed to create PlaylistConfig. Error: %@", error.localizedDescription);
        return false;
    }
    
    // Create Playlist
    NSArray* assets = @[@"SERIES-8-EPISODE-1",
                        @"SERIES-8-EPISODE-2",
                        @"SERIES-8-EPISODE-3",
                        @"SERIES-8-EPISODE-4",
                        @"SERIES-8-EPISODE-5",
                        @"SERIES-8-EPISODE-6"];
    
    
    error = nil;
    self.playlist = [VirtuosoPlaylist create:playlistConfig withAssets:assets error:&error];
    if (error)
    {
        VLog(kVL_LogError, @"Failed to create Playlist. Error: %@", error.localizedDescription);
        return false;
    }
    
    [self loadData];
    return TRUE;
}

-(void)createInitialAsset
{
    VirtuosoAssetConfig* config = [[VirtuosoAssetConfig alloc]initWithURL:@"http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep1/index.m3u8"
                                                                  assetID:@"SERIES-8-EPISODE-1"
                                                              description:@"Sample Description"
                                                                     type:kVDE_AssetTypeHLS];
    if (!config) {
        NSLog(@"VirtuosoAssetConfig create failed.");
        return;
    }
    
    // Configure asset for FastPlay
    config.fastPlayEnabled = true;
    config.offlinePlayEnabled = false;
    config.autoAddToQueue = false;
    
    // Create the asset
    [VirtuosoAsset assetWithConfig:config];
}


- (void)playAsset:(VirtuosoAsset*)asset
{
    [VirtuosoAsset isPlayable:asset callback:^(Boolean playable) {
        if( playable )
        {
            if( asset.type != kVDE_AssetTypeHSS )
            {
                UIViewController* player = [DemoPlayerViewController new];
                
                //
                // Playback FastPlay
                //
                [asset playUsingPlaybackType:kVDE_AssetPlaybackTypeFastPlay
                                   andPlayer:(id<VirtuosoPlayer>)player
                                   onSuccess:^ {
                                       // Present the player
                                       [self presentViewController:player animated:YES completion:nil];
                                   }
                                      onFail:^ {
                                          NSLog(@"Can't play video!");
                                      }];
            }
        }
    }];
}

-(void)loadData
{
    [self.loadQueue addOperationWithBlock:^{
        [self internalLoadData];
    }];
}

- (void)internalLoadData
{
    //
    // Invoked on background thread to prevent blocking UI updates
    //
    if( [[VirtuosoDownloadEngine instance]started] )
    {
        NSArray* assets = [VirtuosoAsset assetsWithAvailabilityFilter:YES];
        
        dispatch_async(dispatch_get_main_queue(), ^{
            [self.assets removeAllObjects];
            [self.assets addObjectsFromArray:assets];
            [self.assets sortUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
                VirtuosoAsset* asset1 = (VirtuosoAsset*)obj1;
                VirtuosoAsset* asset2 = (VirtuosoAsset*)obj2;
                return [asset1.assetID compare:asset2.assetID];
            }];
            
            [self.tableView reloadData];
            [self refreshView];
        });
    }
    else {
        dispatch_after(dispatch_time(DISPATCH_TIME_NOW, 5 * NSEC_PER_SEC), dispatch_get_main_queue(), ^{
            // Try again in 5 seconds
            [self loadData];
        });
    }
}

//
// MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
//

// ------------------------------------------------------------------------------------------------------------
//  Called whenever the Engine starts downloading a VirtuosoAsset object.
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidStartDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self loadData];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports progress for a VirtuosoAsset object
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedForAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self loadData];
}

// ------------------------------------------------------------------------------------------------------------
// Called when an asset is being processed after background transfer
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineProgressUpdatedProcessingForAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self loadData];
}

// ------------------------------------------------------------------------------------------------------------
// Called whenever the Engine reports a VirtuosoAsset as complete
// ------------------------------------------------------------------------------------------------------------
- (void)downloadEngineDidFinishDownloadingAsset:(VirtuosoAsset * _Nonnull)asset
{
    [self loadData];
}

- (void)downloadEngineDidEncounterErrorForAsset:(VirtuosoAsset *)asset error:(NSError *)error task:(NSURLSessionTask *)task data:(NSData *)data statusCode:(NSNumber *)statusCode
{
    [self loadData];
}

- (void)downloadEngineInternalQueueUpdate {
    [self loadData];
}

// ------------------------------------------------------------------------------------------------------------
// Called when Engine start is complete
// ------------------------------------------------------------------------------------------------------------
-(void)downloadEngineStartupComplete:(Boolean)succeeded
{
    [self loadData];
}

-(void)downloadEngineAssetsDeleted:(NSArray<NSString *> *)deletedAssetIDs
{
    [self loadData];
}

-(void)downloadEngineAllAssetsDeleted
{
    self.assetsDeleted = true;
    [self loadData];
}

-(void)downloadEngineDeletedAssetId:(NSString *)assetID
{
    [self loadData];
}

-(void)playlistChange:(VirtuosoPlaylist *)playlist
{
    [self loadData];
}

-(void)downloadEngineFastPlayAssetReady:(VirtuosoAsset *)asset {
    [self loadData];
}
@end

File: Tutorials/Swift/Example7.3/Example7.3/ViewController.swift
//
//  ViewController.swift
//  Example7.1
//
//  Created by Penthera on 3/3/21.
//

import UIKit
import VirtuosoClientDownloadEngine

class ViewController: UITableViewController {
    
    let PlaylistName = "TESTQUEUE-8"
    let MaxItems = 3;
    let InitialAssetList = ["SERIES-8-EPISODE-1", "SERIES-8-EPISODE-2", "SERIES-8-EPISODE-3", "SERIES-8-EPISODE-4", "SERIES-8-EPISODE-5", "SERIES-8-EPISODE-6"]
    
    let backplaneUrl = "https://qa.penthera.com"
    let publicKey = "c9adba5e6ceeed7d7a5bfc9ac24197971bbb4b2c34813dd5c674061a961a899e"
    let privateKey = "41cc269275e04dcb4f2527b0af6e0ea11d227319fa743e4364255d07d7ed2830"
    let loadQueue = OperationQueue()
    
    var assets = [VirtuosoAsset]();
    var playlist : VirtuosoPlaylist?
    var player : DemoPlayerViewController?
    var assetsDeleted = false
    
    var notificationManager : VirtuosoDownloadEngineNotificationManager?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        loadQueue.maxConcurrentOperationCount = 1
        loadQueue.qualityOfService = .userInitiated
        
        
        self.tableView.rowHeight = UITableView.automaticDimension
        self.tableView.estimatedRowHeight = 80
        self.title = "FastPlay Playlist Episodes"
        self.navigationItem.leftBarButtonItem = UIBarButtonItem(title: "Reset", style: .done, target: self, action: #selector(resetClicked))
        self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Browse", style: .done, target: self, action: #selector(addClicked))

        self.notificationManager = VirtuosoDownloadEngineNotificationManager(delegate: self)
        self.startup()
    }

    // MARK: -
    // MARK: Click Handlers
    
    @objc func resetClicked() {
        DispatchQueue.global(qos: .utility).async {
            self.clearData()
        }

    }
    
    @objc func addClicked()
    {
        guard let playlist = self.playlist else { return }
        let sb = UIStoryboard.init(name: "Main", bundle: nil)
        let view = sb.instantiateViewController(identifier: "PlaylistViewController") as PlaylistViewController
        view.playlist = playlist
        self.navigationController?.pushViewController(view, animated: true)
    }

    // MARK: -
    // MARK: Internal Methods
    
    func clearData()
    {
        // Blow away existing Playlist, this needs to happen
        // before we remove the assets.
        VirtuosoPlaylist.clear(self.PlaylistName)
        
        self.playlist = nil
        
        self.assetsDeleted = false;
        // Blow away existing Assets
        VirtuosoAsset.deleteAll()
        
        // wait for all assets to be deleted before
        // adding the first asset.
        while(!self.assetsDeleted) {
            Thread.sleep(forTimeInterval: 1)
        }
        
        // Create demo Playlist
        if (!self.initializeData()) {
            DispatchQueue.main.async {
                let alert = UIAlertController(title: "Initialization error", message: "Check logs for reason.", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                }))
                self.present(alert, animated: true, completion: nil)
            }
            return

        }
        // create initial asset
        if VirtuosoDownloadEngine.instance().started {
            self.createInitialAsset()
        }
        self.loadData()
    }
    
    func refreshView()
    {
        let items = self.playlist?.items.count ?? 0;
        if (items < 6)
        {
            self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Add", style: .done, target: self, action: #selector(addClicked))

        }
        else
        {
            self.navigationItem.rightBarButtonItem = UIBarButtonItem(title: "Browse", style: .done, target: self, action: #selector(addClicked))
        }
    }
    
    
    func startup()
    {
        DispatchQueue.global(qos: .utility).async {
            // Important
            // Each time the app starts we blow away all of the previous demo data.
            // Definitely do NOT want to do this in a production app.
            self.clearData()
            //
            // Enable the Engine
            VirtuosoDownloadEngine.instance().enabled = true;
            
            // Backplane permissions require a unique user-id for the full range of captabilities support to work
            // Production code that needs this will need a unique customer ID.
            // For demonstation purposes only, we use the device name
            let userName = UIDevice.current.name
            
            //
            // Create the engine confuration
            guard let config = VirtuosoEngineConfig(user: userName,
                                                    backplaneUrl: self.backplaneUrl,
                                                    publicKey: self.publicKey,
                                                    privateKey: self.privateKey)
            else
            {
                DispatchQueue.main.async {
                    let alert = UIAlertController(title: "Setup Required", message: "Please contact support@penthera.com to setup the backplaneUrl, publicKey, and privateKey", preferredStyle: .alert)
                    alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                        exit(0)
                    }))
                    self.present(alert, animated: true, completion: nil)
                }
                return
            }

            //
            // Start the Engine
            // This method will execute async, the callback will happen on the main-thread.
            VirtuosoDownloadEngine.instance().startup(config) { (status) in
                if status == .vde_EngineStartupSuccess {
                    print("start succeeded")
                    self.createInitialAsset()
                } else {
                    print("start failed")
                }
            }
        }
    }
    
    func loadData() {
        if (loadQueue.operationCount > 0) {
            loadQueue.cancelAllOperations()
        }
        loadQueue.addOperation {
            self.internalLoadData()
        }
    }
    
    func internalLoadData() {
        
        guard let playlist = self.playlist else {
            return
        }
        let name = playlist.name
        
        // Fetch existing assets
        let assets = VirtuosoAsset.assets(withAvailabilityFilter: true)
        
        // Refresh the Playlst
        let refresh = VirtuosoPlaylist.find(name)
        
        DispatchQueue.main.async {
            self.playlist = refresh
            self.assets.removeAll()
            self.assets.append(contentsOf: assets as! [VirtuosoAsset])
            
            self.assets = self.assets.sorted(by: { (asset1, asset2) -> Bool in
                return asset1.assetID < asset2.assetID
            })
            self.tableView.reloadData()
            
            self.refreshView()
        }
    }
    
    func initializeData() -> Bool
    {
        do {
            
            let config = try VirtuosoPlaylistConfig(name: PlaylistName, playlistType: .vde_PlaylistType_FastPlay)
                             
            
            self.playlist = try VirtuosoPlaylist.create(config, withAssets: InitialAssetList)
            return true
        } catch let error as NSError {
            print("Playlist create failed \(error.localizedDescription)")
            return false
        }
    }

    func createInitialAsset()
    {
        guard let assetConfig = VirtuosoAssetConfig(url: "http://virtuoso-demo-content.s3.amazonaws.com/bbb/season1/ep1/index.m3u8",
                                              assetID: "SERIES-8-EPISODE-1",
                                              description: "Sample Description",
                                              type: .vde_AssetTypeHLS) else {
            return;
        }
        
        // Configure asset for FastPlay
        assetConfig.fastPlayEnabled = true
        assetConfig.offlinePlayEnabled = false
        assetConfig.autoAddToQueue = false
        let _ = VirtuosoAsset.init(config: assetConfig)
    }
    
    func play(asset: VirtuosoAsset) {
        
        VirtuosoAsset.isPlayable(asset) { (playable) in
            if !playable { return }

            switch(asset.type) {
            case .vde_AssetTypeHLS, .vde_AssetTypeDASH, .vde_AssetTypeNonSegmented:
                let demoPlayer = DemoPlayerViewController()
                
                //
                // Playback FastPlay
                //
                asset.play(using: .vde_AssetPlaybackTypeFastPlay, andPlayer: demoPlayer as VirtuosoPlayer, onSuccess: {
                    self.present(demoPlayer, animated: true, completion: nil)
                    
                }) {
                    print("Video is unplayble.")
                }
                break
                
            case .vde_AssetTypeHSS:
                print("Playback of HSS supported in another Tutorial")
                break
                
            @unknown default:
                break;
            }
        }

    }
}

// MARK: -
// MARK: TableView Delegate Methods

extension ViewController 
{
    override func numberOfSections(in tableView: UITableView) -> Int {
        return 1;
    }
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.assets.count;
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let dateformatter = DateFormatter()
        dateformatter.dateStyle = DateFormatter.Style.short
        dateformatter.timeStyle = DateFormatter.Style.short
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "defaultCell", for: indexPath)
        
        if (assets.count <= indexPath.row) {
            cell.textLabel?.text = "Unknown"
            return cell
        }
        
        let asset = self.assets[indexPath.row]
        cell.textLabel?.text  = asset.assetID
        var text = ""
        if (asset.downloadCompleteDateTime != nil) {
            text += "Downloaded: \(dateformatter.string(from: asset.downloadCompleteDateTime!))\n"
        }
        if (asset.fastPlayEnabled) {
            text += "FastPlay Enabled\n"
        }
        if (asset.fastPlayReady) {
            text += "FastPlay Ready\n"
        }

        cell.detailTextLabel?.text = text
        
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        if (assets.count <= indexPath.row) {
            return;
        }
        
        let asset = self.assets[indexPath.row];
        self.play(asset: asset)
        tableView.deselectRow(at: indexPath, animated: true)
    }
}

// MARK: -
// MARK: VirtuosoDownloadEngineNotificationsDelegate Methods


extension ViewController : VirtuosoDownloadEngineNotificationsDelegate {
    func downloadEngineDidStartDownloadingAsset(_ asset: VirtuosoAsset) {
        self.loadData()
    }
    
    func downloadEngineProgressUpdated(for asset: VirtuosoAsset) {
        self.loadData()
    }
    
    func downloadEngineProgressUpdatedProcessing(for asset: VirtuosoAsset) {
        self.loadData()
    }
    
    func downloadEngineDidFinishDownloadingAsset(_ asset: VirtuosoAsset) {
        self.loadData()
    }
    
    func downloadEngineAllAssetsDeleted() {
        self.assetsDeleted = true;
        self.loadData()
    }
    
    func downloadEngineDeletedAssetId(_ assetID: String) {
        self.loadData()
    }
    
    func playlistChange(_ playlist: VirtuosoPlaylist) {
        self.loadData()
    }
    
}
